### ISSUE TITLE
Separating private from public packages

### ISSUE THREAD

**Issue by kevinbosman** [fsprojects/Paket#1517]
I am attempting to use dependency groups to separate our private internal packages (private feed) from public packages (nuget.org).

We want our private packages to always reference the latest version, and we don't want evidence of our private packages leaking out to nuget.org, so we need to specify a single source for our Private group. Using a Private group will hopefully allow us to `paket update group Private` on every build to keep our private packages 100% up to date, without having to update public packages as frequently.

Our PrivatePackage has a dependency on EasyNetQ ~> 0.54.3, which in turn has a dependency on RabbitMQ.Client >= 3.6.0

In the solution, there is a project that explicitly depends on PrivatePackage and RabbitMQ.Client pinned at v3.6.0

paket.references:
```
RabbitMQ.Client

group Private
PrivatePackage
```
paket.dependencies:
```
source https://www.nuget.org/api/v2
nuget RabbitMQ.Client 3.6.0

group Private
source http://localhost:8080/nuget/private
nuget PrivatePackage
```
`paket install` gives the following error:
```
Paket version 2.52.7.0
Resolving packages for group Main:
 - RabbitMQ.Client is pinned to 3.6.0
Resolving packages for group Private:
 - PrivatePackage 1.0.0.0
 - EasyNetQ 0.54.3.424
 - RabbitMQ.Client 3.6.1
Locked version resolution written to paket.lock
Paket failed with:
        Package RabbitMQ.Client is referenced in different versions in ClassLibrary1\ClassLibrary1.csproj
```
Ideally this should resolve to RabbitMQ.Client 3.6.0.

What does "referenced in different versions" mean? From your code it looks like it should read "referenced from different groups"? It would also be helpful to list which groups are referencing the package, because it's tedious to try figure that out when installing hundreds of packages, especially when the one is a transitive dependency it's not obvious at all.

*(Side question):* Is there any way to prevent the dependencies of PrivatePackage from attempting to also resolve from the Private source? These are (mostly) public dependencies and should ideally resolve from nuget.org only. We'd prefer our private feed to only host private packages.

Anyway, if I instead change the RabbitMQ dependency in paket.dependencies to attempt to match the dependency in the PrivatePackage:
```
nuget RabbitMQ.Client ~> 3.6.0
```
then I get the following error from `paket install`:
```
Resolving packages for group Main:
 - RabbitMQ.Client 3.6.1
Resolving packages for group Private:
 - PrivatePackage 1.0.0.0
 - EasyNetQ 0.54.3.424
 - RabbitMQ.Client 3.6.1
Locked version resolution written to paket.lock
Paket failed with:
        The NuGet source https://www.nuget.org/api/v2 for package RabbitMQ.Client was not found in the paket.dependencies file
```
and it generates the following `paket.lock`:
```
NUGET
  remote: https://www.nuget.org/api/v2
  specs:
    RabbitMQ.Client (3.6.1)

GROUP Private
NUGET
  remote: http://localhost:8080/nuget/private
  specs:
    EasyNetQ (0.54.3.424)
      RabbitMQ.Client (>= 3.6.0)
    PrivatePackage (1.0.0.0)
      EasyNetQ (>= 0.54.3 < 0.55.0)
    RabbitMQ.Client (3.6.1)
```

This looks like a possible bug (or possibly just misleading error message), but I'm really trying to find a Paket configuration that allows us to separate our private packages from public ones, and still efficiently manage transitive dependencies.

**Comment by forki** [comment]
Groups are resolved independently from each other. The feature is not intended for such situations.

**Comment by kevinbosman** [comment]
Ok, so is there any other way to achieve this?

**Comment by forki** [comment]
Such things are usually already implemented by local nuget servers. I mean
you only specified one feed in the private group and it still resolved the
public nuget. That is only possible because of nuget.org mirroring or
forwarding of your local nuget server.

**Comment by kevinbosman** [comment]
Sure, that's how I've currently got it set up, as a mirror. But it would be great if I could manage the consistent updates of our private packages independently of public updates, and an added bonus is visually separating the private packages from others in paket.dependencies and paket.references files, so it's obvious which are private vs public.

Is it then not really a supported scenario having 2 (or more) groups in a single paket.references file? It seems it only works flawlessly when there is zero overlap between the dependencies of the 2 groups.

**Comment by forki** [comment]
yes groups are currently designed to be resolved completely independently. That was the design goal.
I can see that there are situations where we want something like groups that get resolved together. I need a moment to think about that.
